tests/unit/test_charm.py

```diff
@@ -1,7 +1,251 @@
-# Copyright 2023 Canonical
+# Copyright 2023 Canonical Ltd.
 # See LICENSE file for licensing details.
 
 
-class TestCharm:
-    def test_dummy_pass(self):
-        assert True
+import tempfile
+import unittest
+from pathlib import Path
+from unittest.mock import Mock, patch
+
+from ops.model import ActiveStatus, BlockedStatus, WaitingStatus
+from scenario import Container, Context, Model, Mount, Relation, State  # type: ignore[import]
+
+from charm import NSSFOperatorCharm
+
+
+@patch("charm.KubernetesServicePatch", new=Mock)
+class TestCharm(unittest.TestCase):
+    def setUp(self):
+        self.ctx = Context(NSSFOperatorCharm)
+
+    def test_given_fiveg_nrf_relation_not_created_when_pebble_ready_then_status_is_blocked(
+        self,
+    ):
+        container = Container(name="nssf", can_connect=True)
+        state_in = State(
+            containers=[
+                container,
+            ]
+        )
+
+        state_out = self.ctx.run(container.pebble_ready_event, state_in)
+
+        self.assertEqual(
+            state_out.status.unit,
+            BlockedStatus("Waiting for fiveg_nrf relation"),
+        )
+
+    def test_given_nrf_data_not_available_when_pebble_ready_then_status_is_waiting(
+        self,
+    ):
+        container = Container(name="nssf", can_connect=True)
+        nrf_relation = Relation("fiveg_nrf")
+        state_in = State(
+            containers=[container],
+            relations=[nrf_relation],
+        )
+
+        state_out = self.ctx.run(container.pebble_ready_event, state_in)
+
+        self.assertEqual(
+            state_out.status.unit,
+            WaitingStatus("Waiting for NRF data to be available"),
+        )
+        self.assertEqual(
+            state_out.deferred[0].name,
+            "nssf_pebble_ready",
+        )
+
+    def test_given_relations_created_and_nrf_data_available_and_storage_not_attached_when_pebble_ready_then_status_is_waiting(  # noqa: E501
+        self,
+    ):
+        container = Container(name="nssf", can_connect=True)
+        nrf_relation = Relation(
+            endpoint="fiveg_nrf",
+            remote_app_name="remote",
+            remote_app_data={"url": "http://nrf:8081"},
+        )
```

Comment:
How about extracting this `nrf_relation` to avoid some code duplication?

Reply:
on the `.replace` comment I added elsewhere: you can do that with the State as well.
in `setUp`:

```python
self.state = State(containers=[self.container], leader=True, relations=[self.nrf_relation])
```

downside to that is that if you need to make a change to the container, you also need to modify the state: in the test:
```python
state = self.state.replace(relations=[self.relation.replace(...))
```

which is ugly. _I think_ I added `State.with_relations(...)` and `with_containers` to catch these common patterns though.

Reply:
I moved the complete relation to `setUp`, and create an empty relation in the only test requiring it.

@PietroPasotti, I checked for `with_relations` and `with_containers`, and it is not in the project yet. I don't dislike having the `State` explicitly defined in each test, it makes the `arrange` portion of the test clear. But redefining the same container, relations can get tiresome.